File: server_elixir/lib/afterlight/activities/snowboard/course.ex
defmodule Afterlight.Activities.Snowboard.Course do
  @moduledoc """
  Canonical Summit Run course — ALPINE RUSH loader, validator and samplers
  (integrate-ssxtricky-snowboard 3.2, ported from the frozen SSXTricky source
  rev e87f6c7d; see the change's baseline.md for provenance).

  The course is ANALYTIC: the browser and this module compute the SAME
  closed-form terrain functions (winding centerline, ground height, ramp
  profiles) that `shared/snowboard/course.js` ports verbatim, while the
  enumerable features (13 ramps, 13 speed zones, 22 pickups, 4 banners) come
  from the SAME canonical document (`priv/snowboard_course.json`,
  byte-identical to `shared/snowboard/course-alpine-rush.json`) with its
  embedded sha256. `tests/fixtures/snowboard/course-parity.json` pins sampled
  golden points; reducer parity (≤1cm position / 0.01 m/s over the golden
  fixture) builds on these agreeing to ≤1e-9.

  There is no baked height grid and no rigid-body dependency: the analytic
  surface IS the mountain for simulation and render alike, exactly as in the
  source game (cosmetic bank noise beyond |x-center| > 29 is render-only and
  absent from contact math on both runtimes).
  """

  defstruct [:doc, :hash, :ramps, :speed_zones, :pickups, :banners, :finish]

  @course_id "alpine-rush"
  @course_version 2
  @rules_version 2
  @length_meters 1800
  @corridor_half_width 35
  @carve_half_width 20
  @groomed_half_width 23
  @edge_bleed_half_width 23
  @ramp_count 13
  @pickup_count 22
  @approx 1.0e-6

  def default_path, do: Application.app_dir(:afterlight, "priv/snowboard_course.json")

  @doc """
  Loads and validates the course document (a decoded JSON map). Raises on
  any contract violation: the race authority must never simulate on top of a
  drifted course. `raise: false` returns `{:error, problems}` instead.
  """
  def load(doc, opts \\ [])

  def load(doc, opts) when is_map(doc) do
    case validate(doc) do
      [] ->
        %__MODULE__{
          doc: doc,
          hash: Map.get(doc, "hash"),
          ramps: Map.get(doc, "ramps"),
          speed_zones: Map.get(doc, "speedZones"),
          pickups: Map.get(doc, "pickups"),
          banners: Map.get(doc, "banners"),
          finish: Map.get(doc, "finish")
        }

      problems ->
        if opts[:raise] == false do
          {:error, problems}
        else
          raise ArgumentError, "refusing to load invalid snowboard course: " <> Enum.join(problems, "; ")
        end
    end
  end

  def load(_other, _opts), do: {:error, ["course document must be an object"]}

  @doc "Total course length in meters."
  def total_length(%__MODULE__{} = course), do: Map.get(course.doc, "lengthMeters")

  @doc "Loads the committed priv copy."
  def load_default do
    default_path() |> File.read!() |> Jason.decode!() |> load()
  end

  @doc """
  Canonical (sorted-key, hash-free) JSON encoding, matching
  `canonicalCourseJson` in shared/snowboard/courseHash.js byte for byte.
  """
  def canonical_json(value) when is_map(value) do
    inner =
      value
      |> Enum.map(fn {k, v} -> {to_string(k), canonical_json(v)} end)
      |> Enum.sort_by(fn {k, _v} -> k end)
      |> Enum.map_join(",", fn {k, v} -> Jason.encode!(k) <> ":" <> v end)

    "{" <> inner <> "}"
  end

  def canonical_json(value) when is_list(value) do
    "[" <> Enum.map_join(value, ",", &canonical_json/1) <> "]"
  end

  def canonical_json(value), do: Jason.encode!(value)

  @doc "sha256 hex over the canonical hash-free form, as computed by the JS export."
  def hash(%{} = doc) do
    without_hash = Map.drop(doc, ["hash", :hash])
    Base.encode16(:crypto.hash(:sha256, canonical_json(without_hash)), case: :lower)
  end

  # --- analytic terrain (verbatim ports of the source functions) -----------------

  @doc """
  Winding groomed centerline x(d). Source:
  sin(d*.003)*24 + sin(d*.009)*7.
  """
  def center_at(d) do
    :math.sin(d * 0.003) * 24 + :math.sin(d * 0.009) * 7
  end

  @doc """
  Absolute ground height. Source:
  -d*.18 + sin(d*.012)*2 + pow(max(0,|x-center(d)|-22),1.18)*.38
  """
  def ground_at(x, d) do
    -d * 0.18 + :math.sin(d * 0.012) * 2 +
      :math.pow(max(0.0, abs(x - center_at(d)) - 22), 1.18) * 0.38
  end

  @doc "Height on a ramp at course distance d (clamped profile)."
  def ramp_height(ramp, d) do
    ramp["base"] + clamp((d - ramp["start"]) / (ramp["end"] - ramp["start"]), 0, 1) * ramp["height"]
  end

  @doc "Whether (x, d) is on the ramp's footprint."
  def on_ramp?(ramp, x, d) do
    d >= ramp["start"] and d <= ramp["end"] and abs(x - ramp["x"]) <= ramp["width"] / 2
  end

  @doc "Contact surface at absolute (x, d): max(ground, ramp surfaces)."
  def surface_at(%__MODULE__{} = course, x, d) do
    Enum.reduce(course.ramps, ground_at(x, d), fn ramp, h ->
      if on_ramp?(ramp, x, d), do: max(h, ramp_height(ramp, d)), else: h
    end)
  end

  # --- validation -----------------------------------------------------------------

  @doc """
  Validates the document and returns a list of problems (empty = valid).
  Mirrors `validateCourse` in shared/snowboard/course.js.
  """
  def validate(doc) when is_map(doc) do
    ramps = fetch(doc, "ramps") || []
    zones = fetch(doc, "speedZones") || []
    pickups = fetch(doc, "pickups") || []
    banners = fetch(doc, "banners") || []

    []
    |> problem(fetch(doc, "id") == @course_id, "course id must be #{@course_id}")
    |> problem(fetch(doc, "version") == @course_version, "course version must be #{@course_version}")
    |> problem(fetch(doc, "rulesVersion") == @rules_version, "rulesVersion must be #{@rules_version}")
    |> problem(fetch(doc, "lengthMeters") == @length_meters, "lengthMeters must be #{@length_meters}")
    |> problem(fetch(doc, "corridorHalfWidth") == @corridor_half_width, "corridorHalfWidth must be #{@corridor_half_width}")
    |> problem(fetch(doc, "carveHalfWidth") == @carve_half_width, "carveHalfWidth must be #{@carve_half_width}")
    |> problem(fetch(doc, "groomedHalfWidth") == @groomed_half_width, "groomedHalfWidth must be #{@groomed_half_width}")
    |> problem(fetch(doc, "edgeBleedHalfWidth") == @edge_bleed_half_width, "edgeBleedHalfWidth must be #{@edge_bleed_half_width}")
    |> problem(length(ramps) == @ramp_count, "course must declare exactly #{@ramp_count} ramps")
    |> problem(ramps_ok?(ramps), "ramps must match the source layout and sit on the terrain")
    |> problem(length(zones) == @ramp_count, "course must declare exactly #{@ramp_count} speed zones")
    |> problem(zones_ok?(zones, ramps), "speed zones must feed their ramps at source offsets")
    |> problem(length(pickups) == @pickup_count, "course must declare exactly #{@pickup_count} pickups")
    |> problem(pickups_ok?(pickups), "pickups must sit on the source lines inside the corridor")
    |> problem(length(banners) == 4, "course must declare the four source banners")
    |> problem(banners_ok?(banners), "the finish banner must sit at the finish")
    |> problem(fetch(fetch(doc, "finish"), "s") == @length_meters, "finish must sit at #{@length_meters}m")
    |> problem(valid_hash?(doc), "hash must match the canonical document")
    |> Enum.reverse()
  end

  def validate(_other), do: ["course document must be an object"]

  defp ramps_ok?(ramps) do
    ramps
    |> Enum.with_index()
    |> Enum.all?(fn {ramp, i} ->
      center = 95 + i * 124
      source_x = center_at(center) + source_offset(i)

      is_binary(fetch(ramp, "id")) and
        abs(fetch(ramp, "start") - (center - 9)) <= @approx and
        abs(fetch(ramp, "end") - (center + 9)) <= @approx and
        abs(fetch(ramp, "x") - source_x) <= 1.0e-4 and
        fetch(ramp, "width") == 12 and fetch(ramp, "height") == 5 and
        abs(fetch(ramp, "base") - ground_at(fetch(ramp, "x"), fetch(ramp, "start"))) <= 1.0e-4 and
        fetch(ramp, "start") > 0 and fetch(ramp, "end") < @length_meters
    end)
  end

  # Source createRamps(): the first ramp sits ON the centerline, the rest
  # alternate (i%3-1)*11 across it.
  defp source_offset(0), do: 0
  defp source_offset(i), do: (rem(i, 3) - 1) * 11

  defp zones_ok?(zones, ramps) do
    zones
    |> Enum.with_index()
    |> Enum.all?(fn {zone, i} ->
      ramp = Enum.at(ramps, i)

      is_map(ramp) and is_map(zone) and
        abs(fetch(zone, "start") - (fetch(ramp, "start") - 41)) <= @approx and
        abs(fetch(zone, "end") - (fetch(ramp, "start") - 19)) <= @approx and
        fetch(zone, "x") == fetch(ramp, "x") and fetch(zone, "width") == 10
    end)
  end

  defp pickups_ok?(pickups) do
    pickups
    |> Enum.with_index()
    |> Enum.all?(fn {pickup, i} ->
      d = 70 + i * 76
      source_x = center_at(d) + :math.sin(i * 2) * 15

      is_map(pickup) and
        abs(fetch(pickup, "d") - d) <= @approx and
        abs(fetch(pickup, "x") - source_x) <= 1.0e-4 and
        abs(fetch(pickup, "x") - center_at(fetch(pickup, "d"))) <= @corridor_half_width and
        fetch(pickup, "d") > 0 and fetch(pickup, "d") < @length_meters
    end)
  end

  defp banners_ok?(banners) do
    Enum.any?(banners, fn banner ->
      is_map(banner) and fetch(banner, "finish") == true and fetch(banner, "d") == @length_meters
    end)
  end

  defp valid_hash?(doc) do
    hash_value = fetch(doc, "hash")
    is_binary(hash_value) and Regex.match?(~r/^[0-9a-f]{64}$/, hash_value) and hash(doc) == hash_value
  end

  defp fetch(map, key) when is_map(map) and is_binary(key), do: Map.get(map, key)
  defp fetch(_other, _key), do: nil

  defp problem(problems, true, _message), do: problems
  defp problem(problems, false, message), do: [message | problems]

  defp clamp(value, low, high), do: value |> max(low * 1.0) |> min(high * 1.0)
end

File: shared/snowboard/course.js
/**
 * Summit Run canonical course — ALPINE RUSH (integrate-ssxtricky-snowboard
 * 3.2). Pure authoring, validation, hashing and sampling for the release
 * route, ported from the frozen user-owned source baseline
 * `SSXTricky/lib/game/rules.mjs` @ rev e87f6c7d (see the change's
 * baseline.md for provenance and file hashes).
 *
 * The course is ANALYTIC, not a baked height grid: the winding centerline,
 * ground height, ramps and speed zones are the source's own closed-form
 * functions, ported verbatim below. The same arithmetic runs in the browser
 * renderer, the client predictor and the Elixir authority
 * (Afterlight.Activities.Snowboard.Course), so render and contact agree
 * exactly inside the corridor. Only the enumerable course features (ramps,
 * speed zones, pickups, banners) are baked into the canonical document so
 * every runtime iterates identical numbers; `x`/`base` values are computed
 * once by the export script and round-trip through JSON exactly.
 *
 * Coordinate convention (source): `d`/`s` downhill meters increases downhill
 * (world z = -s); `x` is the ABSOLUTE world lateral coordinate; the legal
 * corridor is courseCenter(d) ± 35. Source bank noise beyond |x-center| > 29
 * is cosmetic render detail outside the groomed corridor and intentionally
 * absent from contact math, exactly as in the source game.
 */

export const COURSE_ID = 'alpine-rush';
export const COURSE_VERSION = 2;
export const RULES_VERSION = 2;

// Source constants (rules.mjs / engine.js at the frozen revision).
export const LENGTH_METERS = 1800; // source COURSE_LENGTH
export const CORRIDOR_HALF_WIDTH = 35; // source clamp: courseCenter ± 35
export const CARVE_HALF_WIDTH = 20; // source carve window |x-center| < 20
export const GROOMED_HALF_WIDTH = 23; // source groomed display tone boundary
export const BANK_NOISE_START = 29; // source: side > 29 gets bank noise
export const BANK_NOISE_MAX = 18;
export const EDGE_BLEED_HALF_WIDTH = 23; // source: |x-center| > 23 bleeds speed
export const RAMP_COUNT = 13;
export const PICKUP_COUNT = 22;
export const RENDER_SEED = 321; // source terrain/scenery seed (render-side)

const round6 = (value) => (Object.is(value, -0) ? 0 : Math.round(value * 1e6) / 1e6);

// --- source terrain functions (verbatim port) --------------------------------

/** Winding groomed centerline x(d). Source: sin(d*.003)*24 + sin(d*.009)*7. */
export function courseCenter(d) {
  return Math.sin(d * .003) * 24 + Math.sin(d * .009) * 7;
}

/**
 * Absolute ground height. Source:
 * -d*.18 + sin(d*.012)*2 + pow(max(0,|x-center(d)|-22),1.18)*.38
 */
export function groundHeight(x, d) {
  return -d * .18 + Math.sin(d * .012) * 2 + Math.pow(Math.max(0, Math.abs(x - courseCenter(d)) - 22), 1.18) * .38;
}

/** Source createRamps(): 13 ramps, centers 95+i*124, alternating lines. */
export function createRamps() {
  return Array.from({ length: RAMP_COUNT }, (_, i) => {
    const center = 95 + i * 124, x = courseCenter(center) + (i === 0 ? 0 : (i % 3 - 1) * 11);
    return { start: center - 9, end: center + 9, x, width: 12, height: 5, base: groundHeight(x, center - 9) };
  });
}

/** Height on a ramp at course distance d (clamped ramp profile). */
export function rampHeight(r, d) {
  return r.base + clamp((d - r.start) / (r.end - r.start), 0, 1) * r.height;
}

/** Whether (x, d) is on the ramp's footprint. */
export function onRamp(r, x, d) {
  return d >= r.start && d <= r.end && Math.abs(x - r.x) <= r.width / 2;
}

/** Contact surface: max(ground, ramp surfaces) — render/contact agreement. */
export function surfaceHeight(x, d, ramps) {
  return ramps.reduce((h, r) => onRamp(r, x, d) ? Math.max(h, rampHeight(r, d)) : h, groundHeight(x, d));
}

export function clamp(v, lo, hi) {
  return Math.max(lo, Math.min(hi, v));
}

// --- canonical document --------------------------------------------------------

/**
 * The full canonical course document. Deterministic: two calls produce
 * deep-equal structures and the export serializes byte-identically. Ramp and
 * pickup coordinates are baked (computed once, round-tripped exactly through
 * JSON) so JS and Elixir iterate identical numbers.
 */
export function buildCourseDocument() {
  const ramps = createRamps().map((r, i) => ({
    id: `ramp-${i}`,
    start: round6(r.start),
    end: round6(r.end),
    x: round6(r.x),
    width: r.width,
    height: r.height,
    base: round6(r.base),
  }));
  // Source engine: speedZones = ramps.map(r => ({start:r.start-41,end:r.start-19,x:r.x,width:10}))
  const speedZones = ramps.map((r, i) => ({
    id: `zone-${i}`,
    start: round6(r.start - 41),
    end: round6(r.start - 19),
    x: r.x,
    width: 10,
  }));
  // Source engine: 22 pickups at d=70+i*76, x=center(d)+sin(i*2)*15.
  const pickups = Array.from({ length: PICKUP_COUNT }, (_, i) => {
    const d = 70 + i * 76;
    return { id: i, d: round6(d), x: round6(courseCenter(d) + Math.sin(i * 2) * 15) };
  });
  const banners = [
    { d: -12, title: 'ALPINE RUSH', finish: false },
    { d: 600, title: 'GO BIG.', finish: false },
    { d: 1200, title: 'FULL SEND.', finish: false },
    { d: LENGTH_METERS, title: 'FINISH', finish: true },
  ];

  const doc = {
    id: COURSE_ID,
    version: COURSE_VERSION,
    rulesVersion: RULES_VERSION,
    lengthMeters: LENGTH_METERS,
    corridorHalfWidth: CORRIDOR_HALF_WIDTH,
    carveHalfWidth: CARVE_HALF_WIDTH,
    groomedHalfWidth: GROOMED_HALF_WIDTH,
    edgeBleedHalfWidth: EDGE_BLEED_HALF_WIDTH,
    bankNoiseStart: BANK_NOISE_START,
    bankNoiseMax: BANK_NOISE_MAX,
    renderSeed: RENDER_SEED,
    ramps,
    speedZones,
    pickups,
    banners,
    finish: { s: LENGTH_METERS },
  };
  if (!hashImplementation) {
    throw new Error('buildCourseDocument requires setCourseHashImplementation (Node-only authoring)');
  }
  return { ...doc, hash: courseHashOf(doc) };
}

// --- canonical form + hash -----------------------------------------------------
// The BROWSER consumes the document's embedded hash and never recomputes it,
// so node:crypto stays out of this module entirely: Node callers (export
// script, tests) inject the implementation from courseHash.js.
let hashImplementation = null;

/** Node-only hook: inject courseHash from shared/snowboard/courseHash.js. */
export function setCourseHashImplementation(fn) {
  hashImplementation = fn;
}

function courseHashOf(doc) {
  return hashImplementation ? hashImplementation(doc) : null;
}

// --- validation (must pass before any physics use) -------------------------------

const APPROX = 1e-6;

/**
 * Validates one course document and returns a list of problems (empty =
 * valid). Checks identity, source-derived constants, feature invariants and
 * the hash before the rules engine or the renderer ever sample it.
 */
export function validateCourse(doc) {
  const problems = [];
  const at = (ok, message) => { if (!ok) problems.push(message); };

  at(doc && typeof doc === 'object', 'course document must be an object');
  if (!doc || typeof doc !== 'object') return problems;

  at(doc.id === COURSE_ID, `course id must be ${COURSE_ID}`);
  at(doc.version === COURSE_VERSION, `course version must be ${COURSE_VERSION}`);
  at(doc.rulesVersion === RULES_VERSION, `rulesVersion must be ${RULES_VERSION}`);
  at(doc.lengthMeters === LENGTH_METERS, `lengthMeters must be ${LENGTH_METERS}`);
  at(doc.corridorHalfWidth === CORRIDOR_HALF_WIDTH, `corridorHalfWidth must be ${CORRIDOR_HALF_WIDTH}`);
  at(doc.carveHalfWidth === CARVE_HALF_WIDTH, `carveHalfWidth must be ${CARVE_HALF_WIDTH}`);
  at(doc.groomedHalfWidth === GROOMED_HALF_WIDTH, `groomedHalfWidth must be ${GROOMED_HALF_WIDTH}`);
  at(doc.edgeBleedHalfWidth === EDGE_BLEED_HALF_WIDTH, `edgeBleedHalfWidth must be ${EDGE_BLEED_HALF_WIDTH}`);

  // Ramps: the source thirteen, ordered, on the route, bases matching the
  // analytic terrain they sit on.
  const ramps = doc.ramps ?? [];
  at(Array.isArray(ramps) && ramps.length === RAMP_COUNT, `course must declare exactly ${RAMP_COUNT} ramps`);
  const rampIds = new Set();
  ramps.forEach((ramp, i) => {
    at(typeof ramp?.id === 'string' && !rampIds.has(ramp.id), 'ramp ids must be unique strings');
    rampIds.add(ramp?.id);
    const center = 95 + i * 124;
    const sourceX = courseCenter(center) + (i === 0 ? 0 : (i % 3 - 1) * 11);
    at(Math.abs(ramp.start - (center - 9)) <= APPROX, `ramp ${i} start must match the source layout`);
    at(Math.abs(ramp.end - (center + 9)) <= APPROX, `ramp ${i} end must match the source layout`);
    at(Math.abs(ramp.x - sourceX) <= 1e-4, `ramp ${i} x must match the source line`);
    at(ramp.width === 12 && ramp.height === 5, `ramp ${i} must keep the source width/height`);
    at(Math.abs(ramp.base - groundHeight(ramp.x, ramp.start)) <= 1e-4, `ramp ${i} base must sit on the terrain`);
    at(ramp.start > 0 && ramp.end < LENGTH_METERS, `ramp ${i} must sit inside the route`);
  });

  // Speed zones: exactly the source derivation from each ramp.
  const zones = doc.speedZones ?? [];
  at(Array.isArray(zones) && zones.length === RAMP_COUNT, `course must declare exactly ${RAMP_COUNT} speed zones`);
  zones.forEach((zone, i) => {
    const ramp = ramps[i];
    if (!ramp || !zone) return;
    at(Math.abs(zone.start - (ramp.start - 41)) <= APPROX, `zone ${i} must start 41 m before its ramp`);
    at(Math.abs(zone.end - (ramp.start - 19)) <= APPROX, `zone ${i} must end 19 m before its ramp`);
    at(zone.x === ramp.x && zone.width === 10, `zone ${i} must feed its ramp at source width`);
  });

  // Pickups: the source twenty-two, on the route, inside the corridor.
  const pickups = doc.pickups ?? [];
  at(Array.isArray(pickups) && pickups.length === PICKUP_COUNT, `course must declare exactly ${PICKUP_COUNT} pickups`);
  pickups.forEach((p, i) => {
    if (!p) return;
    const d = 70 + i * 76;
    const sourceX = courseCenter(d) + Math.sin(i * 2) * 15;
    at(Math.abs(p.d - d) <= APPROX, `pickup ${i} must sit at the source distance`);
    at(Math.abs(p.x - sourceX) <= 1e-4, `pickup ${i} must sit on the source line`);
    at(Math.abs(p.x - courseCenter(p.d)) <= CORRIDOR_HALF_WIDTH, `pickup ${i} must stay inside the corridor`);
    at(p.d > 0 && p.d < LENGTH_METERS, `pickup ${i} must sit inside the route`);
  });

  // Banners and finish.
  const banners = doc.banners ?? [];
  at(Array.isArray(banners) && banners.length === 4, 'course must declare the four source banners');
  at(banners.some((b) => b?.finish === true && b.d === LENGTH_METERS), 'the finish banner must sit at the finish');
  at(doc.finish?.s === LENGTH_METERS, 'finish must sit at 1800m');

  at(typeof doc.hash === 'string' && /^[0-9a-f]{64}$/.test(doc.hash ?? ''), 'hash must be sha256 hex');
  if (problems.length === 0 && hashImplementation) {
    // Hash verification only where a hash implementation was injected (Node).
    // The browser trusts the server's authoritative copy by construction.
    at(doc.hash === courseHashOf(doc), 'hash must match the canonical document');
  }
  return problems;
}

// --- runtime samplers (identical math to the Elixir authority) -------------------

/**
 * A validated, sampling-ready course. The samplers are the source functions
 * themselves; document features (ramps/zones/pickups) are the baked arrays.
 */
export function loadCourse(doc) {
  const problems = validateCourse(doc);
  if (problems.length > 0) {
    throw new Error(`refusing to load invalid course: ${problems[0]}`);
  }

  /** Absolute-x contact surface including ramps. */
  const surfaceAt = (x, d) => surfaceHeight(x, d, doc.ramps);

  return {
    doc,
    hash: doc.hash,
    lengthMeters: doc.lengthMeters,
    corridorHalfWidth: doc.corridorHalfWidth,
    carveHalfWidth: doc.carveHalfWidth,
    groomedHalfWidth: doc.groomedHalfWidth,
    ramps: doc.ramps,
    speedZones: doc.speedZones,
    pickups: doc.pickups,
    banners: doc.banners,
    finish: doc.finish,
    /** Contact height at absolute (x, d). */
    heightAt: (s, u) => surfaceAt(courseCenter(s) + u, s),
    /** Absolute-x contact height (source convention). */
    surfaceAt,
    centerXAt: (s) => courseCenter(s),
  };
}
